clvm/serialize.py

```diff
@@ -88,24 +182,38 @@ def _op_read_sexp(op_stack, val_stack, f, to_sexp):
         op_stack.append(_op_cons)
         op_stack.append(_op_read_sexp)
         op_stack.append(_op_read_sexp)
-        return
-    val_stack.append(_atom_from_stream(f, b, to_sexp))
+        return val_stack
+    return to_sexp((_atom_from_stream(f, b, to_sexp), val_stack))
 
 
-def _op_cons(op_stack, val_stack, f, to_sexp):
-    right = val_stack.pop()
-    left = val_stack.pop()
-    val_stack.append(to_sexp((left, right)))
+def _op_read_sexp_allow_backrefs(op_stack, val_stack, f, to_sexp):
+    blob = f.read(1)
+    if len(blob) == 0:
+        raise ValueError("bad encoding")
+    b = blob[0]
+    if b == CONS_BOX_MARKER:
+        op_stack.append(_op_cons)
+        op_stack.append(_op_read_sexp_allow_backrefs)
+        op_stack.append(_op_read_sexp_allow_backrefs)
+        return val_stack
+    if b == BACK_REFERENCE:
+        blob = f.read(1)
+        if len(blob) == 0:
+            raise ValueError("bad encoding")
+        path = _atom_from_stream(f, blob[0], lambda x: x)
+        backref = traverse_path(val_stack, path, to_sexp)
```

Comment:
`backref` must be "cloned" (deep copied).
If you modify `backref` in the future, all atoms which refer to the same `backref` will be modified as well.

Reply:
atoms in clvm are immutable though. You're not supposed to change them, nor be allowed to.

Reply:
Yes, SExp should be immutable.
I think immutability should be achieved by implementing getter/setter.
But this is another topic...

Reply:
That's an interesting idea, enforcing immutability using a `setter`. It's been so ingrained in my head that `CLVMObject` items are immutable that I've never even considered mutating one. I'm not aware of anyone trying to do that. But I suppose enforcing it might prevent some foot-shooting.